### PR TITLE
Legolas - Tenants and Resources 

### PR DESCRIPTION
**Pull request by woksin**
## Summary

Implements Contracts 6.3.0, which adds support for getting the configured Tenants, and connection string for a MongoDB resource per tenant. Note: this MongoDB connection string is configured through the "readModels" key in `resources.json` to be backwards compatible with the platform configuration.

### Added

- TenantsService: Getting all tenants through gRPC
- ResourcesService: Getting a specific resource through gRPC. Currently only the MongoDB is supported through the readModels resource configuration

### CODE REVIEW

File: Source/Resources/MongoDB/Configuration.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        /// <summary>
+        /// Gets or sets the database name.
+        /// </summary>
+        public string Database { get; set; }
+
+        /// <summary>
+        /// Gets or sets the value indicating whether or not to use SSL.
+        /// </summary>
+        /// <remarks>This is not used for anything.</remarks>>
+        public bool UseSSL { get; set; }
```
Comment (jakhog) [inline Source/Resources/MongoDB/Configuration.cs:25]: Check if this is used or not

File: Source/Resources/MongoDB/ConnectionDetails.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+using MongoDB.Driver;
+
+namespace Dolittle.Runtime.Resources.MongoDB
+{
+    /// <summary>
+    /// Represents the connection details for a MongoDB resource.
+    /// </summary>
+    /// <param name="ConnectionString">The connection string.</param>
+    /// <param name="DatabaseName">The MongoDB resource database name.</param>
+    public record ConnectionDetails(MongoUrl ConnectionString, string DatabaseName);
```
Comment (jakhog) [inline Source/Resources/MongoDB/ConnectionDetails.cs:13]: Doesn't the MongoUrl already contain the DatabaseName?

File: Source/Resources/MongoDB/IResource.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    /// <summary>
+    /// Defines a system that knows about the MongoDb resource.
+    /// </summary>
+    public interface IResource
+    {
+        /// <summary>
+        /// Gets the <see cref="MongoUrl"/> for the MongoDB resource.
+        /// </summary>
+        /// <returns>The <see cref="MongoUrl"/> connection string.</returns>
+        MongoUrl GetConnectionString();
```
Comment (jakhog) [inline Source/Resources/MongoDB/IResource.cs:17]: Should we be using MongoUrl or ConnectionString here, are they actually the same?

File: Source/Resources/MongoDB/Resource.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    [SingletonPerTenant]
+    public class Resource : IResource
+    {
+        readonly IConfigurationFor<Configuration> _configuration;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Resource"/> class. 
+        /// </summary>
+        /// <param name="configuration">The <see cref="IConfigurationFor{T}"/> <see cref="Configuration"/>.</param>
+        public Resource(IConfigurationFor<Configuration> configuration)
```
Comment (jakhog) [inline Source/Resources/MongoDB/Resource.cs:22]: So we don't need to access the resource type configuration in a special way?

File: Source/Resources/MongoDB/ResourceType.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    /// <summary>
+    /// Represents a <see cref="IAmAResourceType">resource type</see> for the Mongo Db resource.
+    /// </summary>
+    /// <remarks>The name is currently readModels to support for legacy cases.</remarks>
+    public class ResourceType : IAmAResourceType
+    {
+        /// <summary>
+        /// The <see cref="ResourceTypes.ResourceType"/> name.
+        /// </summary>
+        public static ResourceTypes.ResourceType ResourceTypeName => "readModels";
```
Comment (jakhog) [inline Source/Resources/MongoDB/ResourceType.cs:19]: Let's make a remark that this is called "readModels" and not something with MongoDB to be backwards compatible with the old style of resource configuration and use the old databases. Also - let's not this in the release notes so we don't get confused in the future.

File: Source/Resources/MongoDB/ResourceTypeRepresentation.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        static readonly IDictionary<Type, Type> _bindings = new Dictionary<Type, Type>
+        {
+            { typeof(IResource), typeof(Resource) }
+        };
+
+        /// <inheritdoc/>
+        public ResourceTypes.ResourceType Type => ResourceType.ResourceTypeName;
+
+        /// <inheritdoc/>
+        public ResourceTypeImplementation ImplementationName => "n/a";
```
Comment (jakhog) [inline Source/Resources/MongoDB/ResourceTypeRepresentation.cs:22]: Same here, let's make a not of this strange behaviour (and verify that it actually doesn't matter).

File: Source/Resources/MongoDB/Service.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+                return new GetMongoDbResponse(new GetMongoDbResponse
+                {
+                    ConnectionString = _getMongoDbResource().GetConnectionString().ToString()
+                });
+            }
+            catch (Exception ex)
+            {
+                _logger.LogWarning(ex, "Failed to get MongoDb resource for tenant {Tenant}", executionContext.Tenant.Value);
+                return new GetMongoDbResponse {Failure = new Failure(ex.Message) };
+            }
```
Comment (jakhog) [inline Source/Resources/MongoDB/Service.cs:48]: Change loggers to the fancy-fast way of doing it.

File: Source/Server/.dolittle/bounded-context.json [hunk 8]
```diff
@@ -14,10 +14,14 @@
         "embeddings": {
             "production": "MongoDB",
             "development": "MongoDB"
+        },
+        "readModels": {
+            "development": "n/a",
+            "development": "n/a"
```
Comment (jakhog) [inline Source/Server/.dolittle/bounded-context.json:20]: Ahh - this might be why it works. Change this to look more like it does in Production.

Comment (jakhog) [reply Source/Server/.dolittle/bounded-context.json:20]: Also, as codacy says - one of these should be production.

File: Source/Tenancy/LoggerExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    /// <summary>
+    /// Represents a extensions for <see cref="ILogger" />.
+    /// </summary>
+    static class LoggerExtensions
+    {
+        static readonly Action<ILogger, Exception> _failure = LoggerMessage
+            .Define(
+                LogLevel.Warning,
+                new EventId(1231141, nameof(Failure)),
+                "An error occurred");
```
Comment (jakhog) [inline Source/Tenancy/LoggerExtensions.cs:18]: Seems like we can make a little more specific message for what's happening here.

File: Source/Tenancy/Tenancy.csproj [hunk 10]
```diff
@@ -1,16 +1,24 @@
...
+        <PackageReference Include="Microsoft.Extensions.Logging" Version="$(MicrosoftExtensionsVersion)" />
+    </ItemGroup>
+
+    <ItemGroup>
+        <ProjectReference Include="../ApplicationModel/ApplicationModel.csproj" />
+        <ProjectReference Include="../Configuration/Configuration.csproj" />
+        <ProjectReference Include="../Execution/Execution.csproj" />
+        <ProjectReference Include="../Types/Types.csproj" />
+        <ProjectReference Include="../Serialization.Json/Serialization.Json.csproj" />
+        <ProjectReference Include="..\Services\Services.csproj" />
```
Comment (jakhog) [inline Source/Tenancy/Tenancy.csproj:21]: Fix slash direction.

File: versions.props [hunk 11]
```diff
@@ -3,7 +3,7 @@
         <AutofacVersion>5.0.0</AutofacVersion>
         <AutofacExtensionsVersion>6.0.0</AutofacExtensionsVersion>
         <ConsoleTablesVersion>2.4.2</ConsoleTablesVersion>
-        <ContractsVersion>6.2.0</ContractsVersion>
+        <ContractsVersion>6.3.0-legolas.3</ContractsVersion>
```
Comment (jakhog) [inline versions.props:6]: Upgrade to latest.

**Review by jakhog — COMMENTED** [review]
Added some comments for myself to look closer at and try out. Will probably fix or ignore most of these myself :)

File: Source/Resources/MongoDB/ConnectionDetails.cs (hunk 2 above)
Comment (woksin) [reply Source/Resources/MongoDB/ConnectionDetails.cs:13]: It does. I think that record isn't used anymore?

File: Source/Resources/MongoDB/IResource.cs (hunk 3 above)
Comment (woksin) [reply Source/Resources/MongoDB/IResource.cs:17]: The MongoUrl is essentially the connection string

File: Source/Resources/MongoDB/Resource.cs (hunk 4 above)
Comment (woksin) [reply Source/Resources/MongoDB/Resource.cs:22]: I was also thinking that, but I think that is handled bæin the background when getting the IConfogurationFor

File: Source/Tenancy/Tenancy.csproj (hunk 10 above)
Comment (woksin) [reply Source/Tenancy/Tenancy.csproj:21]: Rider does that automatically. Would be nice if we could edit that in Rider

File: versions.props (hunk 11 above)
Comment (jakhog) [reply versions.props:6]: Done

File: Source/Tenancy/Tenancy.csproj (hunk 10 above)
Comment (jakhog) [reply Source/Tenancy/Tenancy.csproj:21]: Done

File: Source/Server/.dolittle/bounded-context.json (hunk 8 above)
Comment (jakhog) [reply Source/Server/.dolittle/bounded-context.json:20]: Done

File: Source/Resources/MongoDB/ResourceTypeRepresentation.cs (hunk 6 above)
Comment (jakhog) [reply Source/Resources/MongoDB/ResourceTypeRepresentation.cs:22]: Done.

File: Source/Resources/MongoDB/Resource.cs (hunk 4 above)
Comment (jakhog) [reply Source/Resources/MongoDB/Resource.cs:22]: Yep - this is the same way we've been doing it for the EventStoreConfiguration :)

File: Source/Resources/MongoDB/ResourceType.cs (hunk 5 above)
Comment (jakhog) [reply Source/Resources/MongoDB/ResourceType.cs:19]: Done

File: Source/Resources/MongoDB/Configuration.cs (hunk 1 above)
Comment (jakhog) [reply Source/Resources/MongoDB/Configuration.cs:25]: It is used, so deleted the remark.

File: Source/Resources/MongoDB/ConnectionDetails.cs (hunk 2 above)
Comment (jakhog) [reply Source/Resources/MongoDB/ConnectionDetails.cs:13]: File wasn't used anymore, so I got rid of it.

File: Source/Resources/MongoDB/IResource.cs (hunk 3 above)
Comment (jakhog) [reply Source/Resources/MongoDB/IResource.cs:17]: Turns out MongoDB uses the MongoUrl under the hood: https://github.com/mongodb/mongo-csharp-driver/blob/30eded69a28ef7be8a01fdc7952dfa66a2fac8d8/src/MongoDB.Driver/MongoClientSettings.cs#L830 so this seems to be the correct thing to use.

File: Source/Tenancy/LoggerExtensions.cs (hunk 9 above)
Comment (jakhog) [reply Source/Tenancy/LoggerExtensions.cs:18]: Done

File: Source/Resources/MongoDB/Service.cs (hunk 7 above)
Comment (jakhog) [reply Source/Resources/MongoDB/Service.cs:48]: Fixed.

**Pull request merged**